app/src/main/java/org/mozilla/fenix/home/sessioncontrol/viewholders/topsites/TopSitesPagerAdapter.kt

```diff
@@ -42,21 +43,43 @@ class TopSitesPagerAdapter(
         }
     }
 
-    private fun update(
+    @VisibleForTesting
+    internal fun update(
         payload: TopSitePagerPayload,
         position: Int,
         adapter: TopSitesAdapter
     ) {
         // Only currently selected page items need to be updated.
-        for (item in payload.changed) {
-            if (item.first < TOP_SITES_PER_PAGE && position == 0) {
-                adapter.notifyItemChanged(item.first, item.second)
-            } else if (item.first >= TOP_SITES_PER_PAGE && position == 1) {
-                adapter.notifyItemChanged(item.first - TOP_SITES_PER_PAGE, item.second)
-            }
+        val currentPageChangedItems = getCurrentPageChanges(payload, position)
+
+        // Build the new list
+        val refreshedItems: MutableList<TopSite> = mutableListOf()
+        refreshedItems.addAll(adapter.currentList)
+
+        // Update new list based on changed items. Mark any removed items for deletion
+        val itemsToRemove: MutableList<Int> = mutableListOf()
+        currentPageChangedItems.forEach { item ->
+            if (item.second.id == -1L)
+                itemsToRemove.add(item.first - (position * TOP_SITES_PER_PAGE))
+            refreshedItems[item.first - (position * TOP_SITES_PER_PAGE)] = item.second
         }
+
+        // Delete any items marked as such and submit the list, then notify adapter of deletions
+        itemsToRemove.forEach { refreshedItems.removeAt(it) }
+        adapter.submitList(refreshedItems)
+        itemsToRemove.forEach { adapter.notifyItemRemoved(it) }
```

Comment:
That's interesting, these two lines are similar. They both ask the adapter to update the UI. Why do we need to do both?

Reply:
What's interesting about this whole method is that we're handling most of the work that DiffUtils should be doing.

Reply:
> That's interesting, these two lines are similar. They both ask the adapter to update the UI. Why do we need to do both?

I think `notifyItemRemoved` was needed before settling on this latest way of using `submitList`. I've removed it and done several tests, and indeed, it's not needed anymore.